lxd/patches.go

```diff
@@ -921,29 +957,48 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 			// Insert storage volumes for snapshots into the
 			// database. Note that snapshots have already been moved
 			// and symlinked above. So no need to do any work here.
-			_, err := dbStoragePoolVolumeCreate(d.db, cs, storagePoolVolumeTypeContainer, poolID, volumeConfig)
-			if err != nil {
-				shared.LogWarnf("Could not insert a storage volume for snapshot \"%s\".", cs)
-				continue
+			_, err := dbStoragePoolVolumeGetTypeID(d.db, cs, storagePoolVolumeTypeContainer, poolID)
+			if err == nil {
+				shared.LogWarnf("Storage volumes database already contains an entry for the container.")
+			} else if err == NoSuchObjectError {
+				// Insert storage volumes for containers into the database.
+				_, err := dbStoragePoolVolumeCreate(d.db, cs, storagePoolVolumeTypeContainer, poolID, volumeConfig)
+				if err != nil {
+					// In case we're on a LXD instance where a a partial
+					// upgrade took place, we should still proceed.
+					shared.LogErrorf("Could not insert a storage volume for snapshot \"%s\".", cs)
+					return err
+				}
+			} else {
+				shared.LogErrorf("Failed to query database: %s", err)
+				return err
 			}
 
 			// Create the new mountpoint for snapshots in the new
 			// storage api.
 			newSnapshotMntPoint := getSnapshotMountPoint(poolName, cs)
-			err = os.MkdirAll(newSnapshotMntPoint, 0711)
-			if err != nil {
-				return err
+			if !shared.PathExists(newSnapshotMntPoint) {
+				err = os.MkdirAll(newSnapshotMntPoint, 0711)
+				if err != nil {
+					shared.LogWarnf("Failed to create mountpoint for snapshot: %s.", newSnapshotMntPoint)
+					failedUpgradeEntities = append(failedUpgradeEntities, fmt.Sprintf("snapshots/%s: Failed to create mountpoint for snapshot.", cs))
+					continue
+				}
 			}
 		}
 
-		os.RemoveAll(snapshotsPath)
+		if shared.PathExists(snapshotsPath) {
+			os.RemoveAll(snapshotsPath)
+		}
```

Comment:
Same comment as above, PathExists isn't necessary if you're not going to check for the error anyway